r/R/dotplot.R

```diff
@@ -1,35 +1,84 @@
+#'               filterBy: Cellsets to subset the experiment with.
+#'               isFilterByAll: Bool value. If true, the experiment will be subsetted to all the cellSets ids.
+#'              }
+#'            }
+#' @param data
+#'
+#' @return
+#' @export
+#'
+#' @examples
 runDotPlot <- function(req, data) {
-  type <- req$body$type
-  req_genes <- req$body$genes
+  useMarkerGenes <- req$body$useMarkerGenes
+  data$dotplot.groups <- NA
+  group_by_cell_sets <- req$body$groupBy$children
+  filter_by <- req$body$filterBy
+  apply_filter <- req$body$applyFilter
 
-  # construct clusters from cell sets
-  data$custom <- NA
-  cells_id <- data$cells_id
-  cell_sets <- req$body$cellSets$children
+  if (length(group_by_cell_sets) < 1) {
+    message("The requested Cell Sets are empty. Returning empty results.")
+    return(list())
+  }
 
-  for (i in seq_along(cell_sets)) {
-    set <- cell_sets[[i]]
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  #Collect ids to subset object
+  if (apply_filter) {
+    subsetIds <- filter_by$cellIds
+  } else {
+    subsetIds <- list()
+    for (child in filter_by$children) {
+      subsetIds <- append(subsetIds, child$cellIds)
+    }
   }
 
-  if (type == "custom") {
-    annot <- data@misc$gene_annotations
-    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
-    annot_subset <- annot_subset[, c("input", "name")]
+  #Subset cell Ids
+  if (length(subsetIds)) {
+    data <- subset_ids(data,subsetIds)
+    cells_id <- data$cells_id
   } else {
-    # type == marker
-    # Need the marker features
+    message("The ids to subset the object are empty. Returning empty results.")
+    return(list())
   }
 
-  dotplot_data <- Seurat::DotPlot(data, features = annot_subset$input, group.by = "custom")$data
-  #features.plot has the ensemble ids
-  dotplot_data$name <- annot_subset[dotplot_data$features.plot, "name"]
+  #Construct the dotplot.groups slot
+  for (i in seq_along(group_by_cell_sets)) {
+    cell_set <- group_by_cell_sets[[i]]
+    filtered_cells <- intersect(cell_set$cellIds, cells_id)
+    #This covers a border case where two cell_sets have the same name (but different ID). Can happen in scratchpad
+    if (cell_set$name %in% data$dotplot.groups) {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- paste0(cell_set$name, i)
+    } else {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- cell_set$name
+    }
+  }
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = dotplot.groups != "NA")
 
-  dotplot_data <- dotplot_data[, c("avg.exp", "pct.exp", "name", "id")]
-  colnames(dotplot_data) <- c("avgExpression", "cellsPercentage", "geneName", "cellSet")
+  #Get marker genes or requested gene names.
+  if (useMarkerGenes) {
+    num_features <- req$body$numberOfMarkers
+    all_markers <- getTopMarkerGenes(num_features, data, group_by_cell_sets)
```

Comment:
I think we want to play around with the parameters to `presto`. Check out [this tutorial](https://divingintogeneticsandgenomics.rbind.io/post/clustered-dotplot-for-single-cell-rnaseq/). Here is the dotplot they generate:

```R
...
markers<- presto::wilcoxauc(pbmc, 'seurat_clusters', assay = 'data')
markers<- presto::top_markers(markers, n = 10, auc_min = 0.5, pct_in_min = 20, pct_out_max = 20)

markers
all_markers<- markers %>%
    select(-rank) %>% 
    unclass() %>% 
    stack() %>%
    pull(values) %>%
    unique() %>%
    .[!is.na(.)]

# Seurats dot plot
p<- DotPlot(object = pbmc, features = all_markers)
p
```
![image](https://user-images.githubusercontent.com/15719520/143152299-b19a9197-865b-44ed-9516-3a6dcfe0dae0.png)

Whereas this is what we get without filtering by `pct.in` and `pct.out`:

```R
markers<- presto::wilcoxauc(pbmc, 'seurat_clusters', assay = 'data')
markers<- presto::top_markers(markers, n = 10)

markers
all_markers<- markers %>%
    select(-rank) %>% 
    unclass() %>% 
    stack() %>%
    pull(values) %>%
    unique() %>%
    .[!is.na(.)]

# Seurats dot plot
p<- DotPlot(object = pbmc, features = all_markers)
p
```
![image](https://user-images.githubusercontent.com/15719520/143152410-fa9e58e9-d8b5-4d5f-b351-2b9f99f106e5.png)

This will also help the heatmap markers